explorer/interpreter/resolve_names.cpp

```diff
@@ -277,11 +279,24 @@ auto NameResolver::ResolveNames(Expression& expression,
       CARBON_ASSIGN_OR_RETURN(std::optional<ValueNodeView> scope,
                               ResolveNames(access.object(), enclosing_scope));
 
-      // If the left-hand side of the `.` is a namespace, resolve the name.
-      // TODO: Look through aliases.
-      if (scope && isa<NamespaceDeclaration>(scope->base())) {
-        auto ns_it =
-            namespace_scopes_.find(&cast<NamespaceDeclaration>(scope->base()));
+      // If the left-hand side of the `.` is a namespace or alias to namespace, resolve the name.
+      if (!scope) {
+        break;
+      }
+
+      std::optional<Nonnull<const NamespaceDeclaration *>> namespace_decl;
+      if (isa<AliasDeclaration>(scope->base())) {
+        const AliasDeclaration& alias = cast<AliasDeclaration>(scope->base());
+        auto alias_it = namespace_aliases_.find(&alias);
+        if (alias_it != namespace_aliases_.end()) {
+          namespace_decl = alias_it->second;
+        }
```

Comment:
Thanks! Can you change this to use `alias.target()` directly here, removing `namespace_aliases_`? That way, we can using the existing tracking instead of adding overlapping storage for aliased namespaces.

Reply:
Do you mean moving `ResolveQualifier()` here?

Can't use `alias.target()` directly because it is of  type `Expression` (not a super type of `NamespaceDeclaration`)?

Reply:
> Do you mean moving `ResolveQualifier()` here?

I wouldn't expect it to, since the alias should already be resolved before reaching this.

> Can't use `alias.target()` directly because it is of type `Expression` (not a super type of `NamespaceDeclaration`)?

You're right, this may involve changes to AliasDeclaration to store its target as an AstNode instead of an Expression. However, it offers greater consistency to store aliased namespaces in the same way as other aliases; aliases are intended to support a wide variety of targets.

Reply:
changing `target` to `AstNode*` doesn't work because we only have `const AstNode*` (ValueNodeView). 

we can't change it to `const AstNode*` because we need `Expression*` from `AliasDeclaration` sometimes.

Added `AliasDeclaration::resolved_declaration`, and found similar pattern is also used in other places. (eg: `ImplDeclaration::match_first`)

Reply:
Thanks, I think I didn't look closely enough. I'm still trying to understand the current code, though. It looks like `resolved_declaration` is going to be `alias.target().base()`. Why doesn't the code in SimpleMemberAccessExpression use `alias.target().base()` directly instead of adding the new `resolved_declaration`?

Reply:
> It looks like resolved_declaration is going to be alias.target().base().

No, alias->target() is just an expression, and resolve_decl is the ast node of declaration of the namespace.

running with 
```
llvm::outs() << "resolved_decl = " << *resolved_decl.value() << " target = " << alias.target() << "\n";
```
resolved_decl = namespace N; target = N